arch/arm/core/cortex_m/mpu/arm_core_mpu.c

```diff
@@ -10,105 +10,295 @@
+		u32_t size = thread->stack_info.size;
+#if !defined(CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT)
+		/* In user-mode the thread stack will include the (optional)
+		 * guard area. For MPUs with arbitrary base address and limit
+		 * it is essential to include this size increase, to avoid
+		 * MPU faults.
+		 */
+		size += thread->stack_info.start - base;
+#endif
+		dynamic_regions[region_num].start = base;
+		dynamic_regions[region_num].size = size;
+		dynamic_regions[region_num].attr = K_MEM_PARTITION_P_RW_U_RW;
+
+		region_num++;
+	}
+#endif /* CONFIG_USERSPACE */
+
+#if defined(CONFIG_MPU_STACK_GUARD)
+	/* Privileged stack guard */
+#if defined(CONFIG_USERSPACE)
+	u32_t guard_start = thread->arch.priv_stack_start ?
+	    (u32_t)thread->arch.priv_stack_start :
+	    (u32_t)thread->stack_obj;
+#else
+	u32_t guard_start = thread->stack_info.start;
+#endif
+	dynamic_regions[region_num].start = guard_start;
+	dynamic_regions[region_num].size = MPU_GUARD_ALIGN_AND_SIZE;
+	dynamic_regions[region_num].attr = K_MEM_PARTITION_P_RO_U_NA;
+
+	region_num++;
+#endif /* CONFIG_MPU_STACK_GUARD */
+
+	/* Configure the dynamic MPU regions */
+	arm_core_mpu_configure_dynamic_mpu_regions(dynamic_regions,
+		region_num);
 }
 
+#if defined(CONFIG_USERSPACE)
+
+/**
+ * @brief Get the maximum number of partitions for a memory domain
+ *        that is supported by the MPU hardware.
+ */
 int _arch_mem_domain_max_partitions_get(void)
 {
 	return arm_core_mpu_get_max_domain_partition_regions();
 }
 
-/*
- * Reset MPU region for a single memory partition
+/**
+ * @brief Configure the memory domain of the thread.
  */
-void _arch_mem_domain_partition_remove(struct k_mem_domain *domain,
-				       u32_t  partition_id)
+void _arch_mem_domain_configure(struct k_thread *thread)
 {
-	ARG_UNUSED(domain);
+	/* Request to configure memory domain for a thread.
+	 * This triggers re-programming of the entire dynamic
+	 * memory map.
+	 */
+	_arch_configure_dynamic_mpu_regions(thread);
+}
 
-	arm_core_mpu_disable();
-	arm_core_mpu_mem_partition_remove(partition_id);
-	arm_core_mpu_enable();
+/*
+ * @brief Reset the MPU configuration related to the memory domain
+ *        partitions
+ *
+ * @param domain pointer to the memory domain (must be valid)
+ */
+void _arch_mem_domain_destroy(struct k_mem_domain *domain)
+{
+	/* This function will reset the access permission configuration
+	 * of the active partitions of the memory domain.
+	 */
+	int i;
+	struct k_mem_partition partition;
+	/* Partitions belonging to the memory domain will be reset
+	 * to default (Privileged RW, Unprivileged NA) permissions.
+	 */
+	k_mem_partition_attr_t reset_attr = K_MEM_PARTITION_P_RW_U_NA;
 
+	for (i = 0; i < CONFIG_MAX_DOMAIN_PARTITIONS; i++) {
+		partition = domain->partitions[i];
+		if (partition.size == 0) {
+			/* Zero size indicates a non-existing
+			 * memory partition.
+			 */
+			continue;
+		}
+		arm_core_mpu_mem_partition_configure(&partition, &reset_attr);
```

Comment:
Thanks for this comment, @pizi-nordic. So, I really thought a lot about this, and I went for this "hybrid" solution of
- re-submitting the map to the driver when we need to actually configure the memory domain on-the-fly (i.e. during current thread's execution)
- just resetting the domain partition / partitions to default ATTRibutes, in case of _destroy or _remove.

The reasons for _not_ doing as you say, are the following:
1. **Performance penalty**: re-submitting the whole map is costly, especially if we only need to remove a single partition (and keep all the remaining ones)
2. We might consider removing this possibility from the Memory Domain API, that is, to be able to _destroy/remove partitions on-the-fly_, and **only** allow it when swapping (discussed with @andrewboie @andyross  and @agross-linaro ).

Reply:
(addressed)